cirq/_compat_test.py

```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
+    imports!
+    """
+    subprocess_context(_test_same_name_submodule_earlier_in_subtree_inner)()
+
+
+def _test_same_name_submodule_earlier_in_subtree_inner():
+    from cirq.testing._compat_test_data.module_a.sub.subsub.dupe import DUPE_CONSTANT
+
+    assert DUPE_CONSTANT
+
+
+def test_metadata_search_path():
+    # to cater for metadata path finders
+    # https://docs.python.org/3/library/importlib.metadata.html#extending-the-search-algorithm
+    subprocess_context(_test_metadata_search_path_inner)()
+
+
+def _test_metadata_search_path_inner():
+    # initialize the DeprecatedModuleFinders
+    # pylint: disable=unused-import
+    import cirq.testing._compat_test_data.module_a
+
+    try:
+        # importlib.metadata for python 3.8+
+        # coverage: ignore
+        import importlib.metadata as m
+    except:  # coverage: ignore
+        # coverage: ignore
+        # importlib_metadata for python <3.8
+        import importlib_metadata as m
+
+    assert m.metadata("flynt")
+
+
+def test_deprecated_module_deadline_validation():
+    with pytest.raises(AssertionError, match="deadline should match vX.Y"):
+        deprecated_submodule(
+            new_module_name="new",
+            old_parent="old_p",
+            old_child="old_ch",
+            deadline="invalid",
+            create_attribute=False,
+        )
+
+
+def test_new_module_is_top_level():
+    subprocess_context(_test_new_module_is_top_level_inner)()
+
+
+def _test_new_module_is_top_level_inner():
+    # sets up the DeprecationFinders
+    # pylint: disable=unused-import
+    import cirq.testing._compat_test_data
+
+    # imports a top level module that was also deprecated
+    from freezegun import api
+
+    assert api
+
+
+def test_loader_failure():
+    class FakeLoader(importlib.abc.Loader):
+        def exec_module(self, module: ModuleType) -> None:
+            raise ValueError("hello")
+
+    with pytest.raises(ValueError, match="hello"):
+        module = types.ModuleType("old")
+        DeprecatedModuleLoader(FakeLoader(), "old", "new").exec_module(module)
+
+
+def test_loader_wrappers():
+    hello_module = types.ModuleType("hello")
+
+    class StubLoader(importlib.abc.Loader):
+        def module_repr(self, module: ModuleType) -> str:
+            return "hello"
+
+        def load_module(self, fullname: str) -> ModuleType:
+            return hello_module
+
+    module = types.ModuleType("old")
+    assert DeprecatedModuleLoader(StubLoader(), "old", "new").module_repr(module) == "hello"
+    assert DeprecatedModuleLoader(StubLoader(), "old", "new").load_module("test") == hello_module
+
+
+def test_invalidate_caches():
+    called = False
+
+    class FakeFinder(importlib.abc.MetaPathFinder):
+        def invalidate_caches(self) -> None:
+            nonlocal called
+            called = True
+
+    DeprecatedModuleFinder(FakeFinder(), "new", None, "old", "v0.1").invalidate_caches()
+    assert called
```

Comment:
Corner case you should also test:
- `cirq_google/file.py` tries to `import cirq.google.something_else`.  (Ideally this will never happen but you probably want to print a sensible error or warning to help developers.)

Reply:
For the first one - as soon as that happens in a pytest, we'll throw an error - we are prohibiting usage of deprecated features from Cirq.

For the second - the `repeated` use cases that I am adding now cover both the old and new access patterns work with a path that has a substring equal to the deprecated module name.